### PR TITLE
Performance Improvements

### PR DESCRIPTION
**Pull request by dleslie**
No longer rely on Apropos for matching. Apropos would perform a slow substring or regex search at every call; as well as rebuilding the entire list of available symbols. Now the list of symbols is cached and a much faster prefix-matching method is employed.

In testing with the #100 test cases I see a 0.5s hitch at first call followed by a 0s+epsilon hit for subsequent calls.

### CODE REVIEW

File: scheme/chicken/geiser/emacs.scm [hunk 1]
```diff
@@ -231,7 +231,25 @@
...
-
+  
+  (define (symbol-information-list)
+    (memoize
+     '(symbol-information-list)
+     (lambda ()
+       (map (lambda (lst)
+	      (let-values (((name module) (remove-internal-name-mangling (car lst))))
+		(append (list name module) (cdr lst))))
+	    (apropos-information-list "" #:macros? #t)))))
```
Comment (dleslie) [inline scheme/chicken/geiser/emacs.scm]: This is a bit of a hack.

```apropos-information-list``` returns the full environment if given an empty string to match. That's an undocumented 'feature'.

File: scheme/chicken/geiser/emacs.scm [hunk 2]
```diff
@@ -369,87 +388,86 @@
...
-
-	  `(,entry-str ("args" (("required" ,@reqs)
-				("optional" ,@opts)
-				("key" ,@keys)))
-		       ("module" ,@(make-module-list sym module))))))))
+  
+  (define (fmt node)
+    (memoize
+     `(fmt ,node)
+     (lambda ()
```
Comment (dleslie) [inline scheme/chicken/geiser/emacs.scm]: Spacing changes caused by memoizing.

File: scheme/chicken/geiser/emacs.scm [hunk 3]
```diff
@@ -369,87 +388,86 @@
...
-       (apropos-information-list sym #:macros? #t)))))
+    (memoize
+     `(find-signatures ,sym)
+     (lambda ()
+       (let ((str (symbol->string sym)))
+	 (map
+	  (cut fmt <>)
+	  (filter
+	   (lambda (lst) (string=? (car lst) str))
+	   (find-symbol-information sym)))))))
```
Comment (dleslie) [inline scheme/chicken/geiser/emacs.scm]: Removing name mangling is pushed down to ```find-symbol-information```.

File: scheme/chicken/geiser/emacs.scm [hunk 4]
```diff
@@ -231,7 +231,25 @@
...
 	    (hash-table-set! table tag (thunk))
 	    (hash-table-ref table tag)))))
-
+  
+  (define (symbol-information-list)
+    (memoize
+     '(symbol-information-list)
+     (lambda ()
+       (map (lambda (lst)
+	      (let-values (((name module) (remove-internal-name-mangling (car lst))))
```
Comment (dleslie) [inline scheme/chicken/geiser/emacs.scm]: We never, ever use the mangled name. So just de-mangle to begin with.

Hilariously, this extra work doesn't push this over the 0.55s that I was seeing when relying on apropos to do all the work. And we only do it once, so long as the environment doesn't change.

**Comment by jaor** [comment]
dan, this looks fine to me... do you want to wait until we know if it helps in #122 or should i merge right away?

**Comment by dleslie** [comment]
Merge in; it helps regardless, I'm curious about #122 but it may not be related to what I changed.

**Pull request merged**